kernel/src/process.rs

```diff
@@ -243,6 +243,52 @@ pub trait ProcessType {
     fn debug_syscall_called(&self);
 }
 
+/// Generic trait for implementing process restart policies.
+///
+/// This policy allows a board to specify how the kernel should decide whether
+/// to restart an app after it crashes.
+pub trait ProcessRestartPolicy {
+    /// Decide whether to restart the `process` or not.
+    ///
+    /// Returns `true` if the process should be restarted, `false` otherwise.
+    fn should_restart(&self, process: &dyn ProcessType) -> bool;
+}
+
+/// Implementation of `ProcessRestartPolicy` that uses a threshold to decide
+/// whether to restart an app. If the app has been restarted more times than the
+/// threshold then the app will no longer be restarted.
+pub struct ThresholdRestart {
+    threshold: usize,
+}
+
+impl ThresholdRestart {
+    pub const fn new(threshold: usize) -> ThresholdRestart {
+        ThresholdRestart { threshold }
+    }
+}
+
+impl ProcessRestartPolicy for ThresholdRestart {
+    fn should_restart(&self, process: &dyn ProcessType) -> bool {
+        process.debug_restart_count() <= self.threshold
```

Comment:
I don't remember the original PR that added the various `debug_x` functions to `ProcessType`, but I have always assumed that the intention of those functions and the associated state being prefixed with "debug_" was to make it easy to pull those out for any particularly constrained deployment. If we are using `debug_restart_count` as part of the restart policy, does it still make sense for that function to be considered a debug function?

Reply:
That's a good point.